### PR TITLE
DO-175: Added error handling framework.

### PR DESCRIPTION
**Pull request by dhruvlahoti**

### CODE REVIEW

File: src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx [hunk 1]
```diff
@@ -109,6 +126,10 @@ class MemberDetails extends React.Component {
         <div key={memberDetails.id}>
           {/* {TODO we need to break down it into components with dynamic values currently it's not configurable}*/}
           <div className="member-details-view-route__header" onClick={this.props.onClickCloseModal}>
+            <Error
+              type={ERROR_TYPES.MEMBER_SCORE}
+              onDismiss={this.onScoreErrorDismiss}
```
Comment (pepcus-pratik) [inline src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx:131]: I believe the error dismiss action should only be called from the `Error` component most of the time. `onDismiss` callback can be kept to use do some operations when an error is closed.

File: src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx [hunk 2]
```diff
@@ -390,7 +411,13 @@ class MemberDetails extends React.Component {
         </div>
       );
     }
-    return null;
+    return (
+      <Error
+        type={ERROR_TYPES.SELECTED_MEMBER}
+        onDismiss={this.onMemberErrorDismiss}
+        style={{ top: '60px' }}
```
Comment (pepcus-pratik) [inline src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx:418]: Will it be good if we do the styling using a class, what is preferred by the CSS team? 
@amitjain24524

File: src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx [hunk 3]
```diff
@@ -402,6 +429,7 @@ MemberDetails.propTypes = {
   getMemberScoreAction: PropTypes.func,
   getMemberDetailsByIdAction: PropTypes.func,
   resetCurrentMember: PropTypes.func,
+  popErrorAction: PropTypes.func,
```
Comment (pepcus-pratik) [inline src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx:432]: Just a comment related to the practice, we should always mark actions as **required**

File: src/components/routes/Member/AuthenticatedMember/MemberProfileViewRoute/MemberProfileEditView/MemberProfileEditView.jsx [hunk 4]
```diff
@@ -147,7 +163,13 @@ class MemberProfileEditView extends React.Component {
         </div>
       );
     }
-    return null;
+    return (
+      <Error
+        type={ERROR_TYPES.SELECTED_MEMBER}
+        onDismiss={this.onMemberErrorDismiss}
+        style={{ top: '60px' }}
```
Comment (pepcus-pratik) [inline src/components/routes/Member/AuthenticatedMember/MemberProfileViewRoute/MemberProfileEditView/MemberProfileEditView.jsx:170]: This `style` passed in Error is too redundant. Can we get rid of it?

File: src/components/routes/Member/AuthenticatedMember/UpdateCredentialsRoute/UpdateCredentialsContent/UpdateCredentialsContent.jsx [hunk 5]
```diff
@@ -18,6 +25,11 @@ class UpdateCredentialsContent extends React.Component {
     };
   }
 
+  componentWillUnmount() {
+    this.props.popErrorAction(ERROR_TYPES.UPDATE_USERNAME);
+    this.props.popErrorAction(ERROR_TYPES.UPDATE_PASSWORD);
```
Comment (pepcus-pratik) [inline src/components/routes/Member/AuthenticatedMember/UpdateCredentialsRoute/UpdateCredentialsContent/UpdateCredentialsContent.jsx:30]: Clearing errors in `componentWillUnmount` will be vulnerable to bugs

File: src/services/errors/reducers/errors.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+          ...state.list,
+          {
+            type: action.errorType,
+            message: action.errorMessage,
+          },
+        ],
+      };
+
+    case 'POP_ERROR': {
+      if ((state.list.length)) {
```
Comment (pepcus-pratik) [inline src/services/errors/reducers/errors.js:20]:
```suggestion
      if (state.list.length) {
```

File: src/services/errors/reducers/errors.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        const updatedList = [];
+        (state.list).forEach((data) => {
```
Comment (pepcus-pratik) [inline src/services/errors/reducers/errors.js:22]: Please use the native `Array.filter` method

File: src/services/member/sagas/member.js [hunk 8]
```diff
@@ -65,13 +67,16 @@ export function* getMembersListSaga(action) {
...
-    }));
+    yield put(resolveSaga(getMemberListSuccessAction,
+      true, ERROR_TYPES.MEMBER_LIST, {
+        ...response,
+      }));
     yield put(setAppLoadingDisabled());
   } catch (e) {
-    console.error('ERROR OCCURRED IN FETCHING MEMBER LIST >>>', e);
-    yield put(getMemberListFailedAction(e));
+    const errorMessage = hasIn(e, 'response.data.error.message') ? e.response.data.error.message
```
Comment (pepcus-pratik) [inline src/services/member/sagas/member.js:76]: Use lodash `get` method instead

File: src/services/middlewares/errorHandling.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { popErrorAction, pushErrorAction } from '../errors/actions/errors';
+
+export const errorHandling = ({ dispatch, getState }) => next => (action) => {
```
Comment (pepcus-pratik) [inline src/services/middlewares/errorHandling.js:3]: Please add JSDoc explaining the purpose, usage, and result of this middleware

File: src/services/middlewares/errorHandling.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+  if (action.handleError) {
+    if ((action.type).includes('SUCCESS')) {
```
Comment (pepcus-pratik) [inline src/services/middlewares/errorHandling.js:5]: Checking **_SUCCESS** and **_FAIL** at end of the string will be a better check

File: src/utils/core/coreUtils.js [hunk 11]
```diff
@@ -140,4 +140,24 @@ export const onSessionExpired = () => {
...
   }
-} ;
+};
+
+/**
+ * Resolve the saga success or failure actions.
+ * @param {Function} action
+ * @param {Boolean} handleError
+ * @param {String} errorType
+ * @param {String || Object} response
```
Comment (pepcus-pratik) [inline src/utils/core/coreUtils.js:150]: It can by `any`

File: src/services/middlewares/errorHandling.js (hunk 10 above)
Comment (dhruvlahoti) [reply src/services/middlewares/errorHandling.js:5]: We have many of the actions which are like FAILED_ACTION or SUCCESS_ACTION, so i think checking SUCCESS & FAIL in the action type will be beneficial, until we make all the action names to the desired format. Also, we are anyway using handleError flag for invocation.

File: src/utils/core/coreUtils.js (hunk 11 above)
Comment (dhruvlahoti) [reply src/utils/core/coreUtils.js]: Done

File: src/services/middlewares/errorHandling.js (hunk 9 above)
Comment (dhruvlahoti) [reply src/services/middlewares/errorHandling.js]: Done

File: src/services/errors/reducers/errors.js (hunk 7 above)
Comment (dhruvlahoti) [reply src/services/errors/reducers/errors.js]: Done

File: src/services/errors/reducers/errors.js (hunk 6 above)
Comment (dhruvlahoti) [reply src/services/errors/reducers/errors.js]: Done

File: src/services/member/sagas/member.js (hunk 8 above)
Comment (dhruvlahoti) [reply src/services/member/sagas/member.js]: Done

File: src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx (hunk 1 above)
Comment (dhruvlahoti) [reply src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx]: Done

File: src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx (hunk 2 above)
Comment (dhruvlahoti) [reply src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx]: DOne

File: src/components/routes/Member/AuthenticatedMember/MemberProfileViewRoute/MemberProfileEditView/MemberProfileEditView.jsx (hunk 4 above)
Comment (dhruvlahoti) [reply src/components/routes/Member/AuthenticatedMember/MemberProfileViewRoute/MemberProfileEditView/MemberProfileEditView.jsx]: Done

File: src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx (hunk 3 above)
Comment (dhruvlahoti) [reply src/components/routes/Member/AuthenticatedMember/MemberDetailsViewRoute/MemberDetailsViewRoute.jsx]: Done

File: src/components/member/MemberDetailsWizard/WizardForm/WizardForm.jsx [hunk 12]
```diff
@@ -166,6 +171,7 @@ class WizardForm extends React.Component {
       this.props.setMemberSuccess(false);
       this.props.onNext();
     }
+    this.onMemberErrorDismiss();
```
[the remark this answers is not in the document]
Comment (dhruvlahoti) [reply src/components/member/MemberDetailsWizard/WizardForm/WizardForm.jsx]: Done

File: src/components/member/MemberDetailsWizard/WizardForm/WizardForm.jsx [hunk 13]
```diff
@@ -313,6 +305,10 @@ class WizardForm extends React.Component {
     return null;
   };
 
+  onMemberErrorDismiss = () => {
```
[the remark this answers is not in the document]
Comment (dhruvlahoti) [reply src/components/member/MemberDetailsWizard/WizardForm/WizardForm.jsx]: Done.

File: src/services/errors/reducers/errors.js [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+  switch (action.type) {
+    case 'PUSH_ERROR':
+      return {
+        ...state,
+        list: [
+          ...state.list,
+          {
+            type: action.errorType,
+            message: action.errorMessage,
+          },
```
[the remark this answers is not in the document]
Comment (dhruvlahoti) [reply src/services/errors/reducers/errors.js:28]: Fixed.

File: src/components/routes/Member/AuthenticatedMember/UpdateCredentialsRoute/UpdateCredentialsContent/UpdateCredentialsContent.jsx (hunk 5 above)
Comment (dhruvlahoti) [reply src/components/routes/Member/AuthenticatedMember/UpdateCredentialsRoute/UpdateCredentialsContent/UpdateCredentialsContent.jsx]: Removed

File: src/services/middlewares/errorHandling.js (hunk 10 above)
Comment (dhruvlahoti) [reply src/services/middlewares/errorHandling.js:12]: For now added Fixme, will change the code once all the actions gets in correct format.

File: src/components/common/Error/Error.jsx [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+} = ErrorsModule.actions;
+
+/** This component is responsible for showing errors we receive from back-end
+ * @param {Object} props
+ * @return {null|*}
+ * @constructor
+ */
+const Error = (props) => {
+  const { errorList, onDismiss, type, className } = props;
+  if (errorList.length) {
```
[the remark this answers is not in the document]
Comment (dhruvlahoti) [reply src/components/common/Error/Error.jsx]: Done

**Comment by pepcus-pratik** [comment]
LGTM 👍

**Comment by mohitsaladiya** [comment]
I think you've missed bootstrap saga. Make sure that every saga got covered in this.

File: src/components/common/Error/Error.jsx [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+/** This component is responsible for showing errors we receive from back-end
+ * @param {Object} props
+ * @return {null|*}
+ * @constructor
+ */
+class Error extends React.Component {
+
+  componentWillUnmount() {
+    this.props.popErrorAction(this.props.type);
```
Comment (mohitsaladiya) [inline src/components/common/Error/Error.jsx:23]: This action should only get called once the errorType is matched in the existing errors in the reducer.

File: src/components/common/Error/Error.jsx [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  render() {
+    const { errorList, onDismiss, type, className } = this.props;
+    if (errorList.length) {
+      let errorMessage = '';
+      errorList.forEach((error) => {
```
Comment (mohitsaladiya) [inline src/components/common/Error/Error.jsx:30]: I think we should create a selector for this say getErrorMessageByType which will return us the required error message by matching the provided errror type.

File: src/components/common/Error/Error.jsx (hunk 16 above)
Comment (dhruvlahoti) [reply src/components/common/Error/Error.jsx]: Done

File: src/components/common/Error/Error.jsx (hunk 17 above)
Comment (dhruvlahoti) [reply src/components/common/Error/Error.jsx]: Done

**Pull request merged**